### PR TITLE
feat(remote_wal): introduce region meta and integrate it into creating table metadata

### PR DESCRIPTION
**Pull request by niebayes**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- Define a `RegionMeta` to store per-region unique metadata. Currently, it stores per-region unique wal metadata which is stored in `RegionWalMeta`.
- Add `RegionMeta` in `DatanodeTableValue` and integrate persisting region metas into creating table metadata.
- Update tests accordingly.

This PR is part of the Kafka Remote Wal project. With this feature, the `TableMetadataAllocator` is able to allocate per-region unique metadata while handling create table requests. The allocated metadata will be persisted during creating table metadata by the procedure manager. On the other hand, a datanode is able to fetch regions' unique metadata during initializing region servers, so that regions can be opened properly.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/pull/2753
https://github.com/GreptimeTeam/greptimedb/pull/2816

### CODE REVIEW

File: src/common/meta/src/key/datanode_table.rs [hunk 1]
```diff
@@ -101,16 +102,25 @@ impl TableMetaKey for DatanodeTableKey {
 pub struct DatanodeTableValue {
     pub table_id: TableId,
     pub regions: Vec<RegionNumber>,
+    /// Region's unique metadata or options.
+    pub region_metas: Vec<RegionMeta>,
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/common/meta/src/key/datanode_table.rs:106]: > 1. Maybe a `HashMap<RegionId, RegionMeta>` with the `#[serde(default)]` can avoid this problem?

Vote for this.

File: src/common/meta/src/region_meta/wal_meta.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            name,
+        }
+    }
+}
+
+impl ToString for RegionWalMetaKey {
+    fn to_string(&self) -> String {
+        format!("{}/{}/{}", self.table_id, self.region_number, self.name)
+    }
+}
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/region_meta/wal_meta.rs:59]: I'm working on removing the `RegionWalMeta` and only keeping the `RegionMeta`.

File: src/common/meta/src/key/datanode_table.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/key/datanode_table.rs:106]: @MichaelScofield On initializing a region server, metadata of a table's regions must be fetched. If the `DatanodeTableValue` does not contain the region metas, we need to invoke one RPC request for each region. These operations are expensive. Therefore, we choose to store a copy of region meta in `DatanodeTableValue` for each region.

[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/key/datanode_table.rs:106]: @WenyXu Changed to use `HashMap<RegionId, RegionMeta`. But, how about use `HashMap<RegionNumber, RegionMeta>` ?

File: src/common/meta/src/region_meta/wal_meta.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/region_meta/wal_meta.rs:59]: @WenyXu The `RegionWalMeta` is now replaced with the `RegionWalOptions` which is of type `HashMap<String, String>`.

File: src/common/meta/src/key/datanode_table.rs [hunk 3]
```diff
@@ -165,15 +175,22 @@ impl DatanodeTableManager {
         engine: &str,
         region_storage_path: &str,
         region_options: HashMap<String, String>,
+        region_meta_map: HashMap<RegionNumber, RegionMeta>,
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/meta/src/key/datanode_table.rs:178]: @WenyXu Let's use `HashMap<RegionNumber, RegionMeta>` as what was mentioned in https://github.com/GreptimeTeam/greptimedb/pull/2830#discussion_r1407461387

File: src/common/meta/src/key/datanode_table.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key/datanode_table.rs:106]: > @WenyXu Changed to use `HashMap<RegionId, RegionMeta`. But, how about use `HashMap<RegionNumber, RegionMeta>` ?

Yep, The `RegionNumebr` is better.

**Pull request closed without merging**